lib/TransferStatusHelpers/SharedTransferData.cs

```diff
@@ -36,23 +58,83 @@ public long TotalLength
         /// </summary>
         public TransferJob TransferJob { get; set; }
 
-        /// <summary>
-        /// Gets or sets list of available transfer data from source.
-        /// </summary>
-        public ConcurrentDictionary<long, TransferData> AvailableData { get; set; }
-
         /// <summary>
         /// Gets or sets a value indicating whether should disable validation of content md5.
         /// The reader should get this value from source's <c>RequestOptions</c>,
         /// the writer should do or not do validation on content md5 according to this value.
         /// </summary>
         public bool DisableContentMD5Validation { get; set; }
-        
+
         /// <summary>
         /// Gets or sets attribute for blob/azure file.
         /// </summary>
         public Attributes Attributes { get; set; }
 
-        public event EventHandler OnTotalLengthChanged;
+        public bool TryAdd(long key, TransferData value)
```

Comment:
Add more comments for public methods (your ideas)! The code looks great to me! Add comments to make it maintainable for other and future engineers.